### ISSUE TITLE
Training labels not applied properly to training data

### ISSUE THREAD

**Issue by TheOfficialOzone** [ultralytics/ultralytics#17796]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

# Bug
Labels are not included in the generated train_batch**X**.jpg images during training of a segmentation model.
Code to reproduce at bottom of section including the example training data.

## Likely cause of bug
I am not familiar with how the training label images are generated, however I highly suspect the issue is that if there are no points that define the polygon (label) in the image. This is caused when Yolo performs augmentation such as crop, resize, stretch, etc as it can morph the label such that all points defining the label are outside the image. This causes the mask to encompress up to the entire image but still not be included
## Examples
- All white parts of the images are included in the label, thus if they are unlabelled the bug has occured
![train_batch41](https://github.com/user-attachments/assets/ff8243c4-badb-4ea9-a5c0-64b9c28fbef6)
![train_batch42](https://github.com/user-attachments/assets/17895e1b-a967-4c6d-8a18-39b59962893d)

### Code to reproduce, instuctions in other section
[GitIssues.zip](https://github.com/user-attachments/files/17916419/GitIssues.zip)

### Environment

```
Ultralytics 8.3.29 🚀 Python-3.10.12 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24564MiB)
Setup complete ✅ (32 CPUs, 15.5 GB RAM, 23.5/251.0 GB disk)

OS                  Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 15.47 GB
Disk                23.5/251.0 GB
CPU                 13th Gen Intel Core(TM) i9-13900
CPU count           32
GPU                 NVIDIA GeForce RTX 4090, 24564MiB
GPU count           1
CUDA                12.1

numpy               ✅ 2.1.2>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 5.4.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1>=1.8.0
torchvision         ✅ 0.19.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.11>=2.0.0
numpy               ✅ 2.1.2<2.0.0; sys_platform == "darwin"
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"
```

### Minimal Reproducible Example

# How to reproduce
1. Download & Extract provided training images, config (.yaml) and test_yolo.py file 
2. Edit .yaml file such that the folder path is correct
3. Run test_yolo.py 
4. Examine the generated train_batch**X**.jpg images to see if the bug occured (You may need to train more than once)

## What to look for
- Any part that is white is labelled, so if any white pixels are unlabelled this bug has occured

### Examples
![train_batch0](https://github.com/user-attachments/assets/fe7f5b3f-1b00-4004-beb1-a50b5d5413b0)
- In this case the bottom left image is clearly white, but unlabelled

![train_batch2](https://github.com/user-attachments/assets/25cd0a90-8e46-48e8-ba99-0d15cf620719)
- Top right image does has white, but it isn't labelled

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TheOfficialOzone, thank you for bringing this to our attention 🚀! We understand that you're encountering an issue with labels not being applied correctly during the training of a segmentation model on the Ultralytics repository.

For us to assist you effectively, please ensure that you've provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) if it's not already included in your report. This will help us understand and address the issue more efficiently. It seems like you've already attached a code example and some test images, which is great!

📄 In the meantime, we suggest ensuring all your dependencies are up-to-date. Upgrade to the latest `ultralytics` package, including all requirements, within a Python >=3.8 environment using PyTorch >=1.8 to see if the issue persists:

```bash
pip install -U ultralytics
```

For further tips and tricks regarding custom training, please refer to our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our Ultralytics community for real-time support or discussions:

- Head over to [Discord](https://ultralytics.com/discord) for chat support 🎧
- Visit [Discourse](https://community.ultralytics.com) for deeper discussions
- Share experiences or get insightful feedback on our [Subreddit](https://reddit.com/r/ultralytics)

Finally, an Ultralytics engineer will review the details of your issue soon and follow up with you for additional help. Thank you for your patience and cooperation!

## Environments

In case you wish to shift your work to a more verified environment, you might consider:

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

For more details on different environments, please refer to the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/).

We appreciate your engagement with the Ultralytics repository and hope to resolve your issue soon! 🌟

**Comment by TheOfficialOzone** [comment]
This issue persists after running `pip install -U ultralytics`.
The version that was upgraded to was ultralytics 8.3.37.

**Comment by Y-T-G** [comment]
@Laughing-q I was debugging this and [this](https://github.com/Y-T-G/ultralytics/commit/070208830f4dc007c290fae776fe4350e37dc79a) change no longer caused the issue.

It's also more similar to what `Instances.clip()` does which is bring the out of bound segments into bounds, which are otherwise dropped by `segment2box`.

**Comment by Laughing-q** [comment]
@Y-T-G Nice work! probably we should introduce the new clip operation you introduced and also keep the two lines:
```bash
        segments[..., 0] = segments[..., 0].clip(bboxes[:, 0:1], bboxes[:, 2:3])
        segments[..., 1] = segments[..., 1].clip(bboxes[:, 1:2], bboxes[:, 3:4])
```
The reason why I added the range restriction of boxes was to ensure the boxes/segments fit objects more for OBB task(we are using the same `xy` format as segment task for obb).

**Comment by Y-T-G** [comment]
@Laughing-q I see. Makes sense. I didn't think of OBB.

**Comment by TheOfficialOzone** [comment]
@Y-T-G Your fix worked on both my example and my actual dataset. Thanks a bunch!

**Comment by Laughing-q** [comment]
@Y-T-G perhaps we could add this additional clip you added inside `segment2box`, might be a better solution for both obb and segment tasks.

**Comment by Laughing-q** [comment]
@Y-T-G Because we better not to directly clip the original segment, or for obb points it might be distorted by this operation and then the obb boxes will be less fit to objects
